### PR TITLE
feat: add LoopDetectionMiddleware to break repetitive tool call loops

### LINKED ISSUE #1055: Agent gets stuck in repetitive tool call loops until recursion limit
## Problem

The agent (both lead agent and subagents) can get stuck calling the same tool with the same arguments repeatedly until `recursion_limit` kills the run. There is no loop detection or repetition tracking.

### Example scenario

When given a task like "audit test imports and write a report", the agent:
1. Calls `bash` to check if a module exists
2. Gets the result
3. Calls `bash` with the exact same command again
4. Repeats steps 2-3 until recursion limit (default 25, configurable to 100+) is hit
5. Run ends with `GraphRecursionError` or `status=error`

This is related to #55 — the recursion limit is a blunt safety net, but the real fix is detecting repetition early and breaking the loop gracefully.

### Root cause

The ReAct loop (`create_agent()`) has no mechanism to:
- Track recent tool calls
- Detect identical calls being repeated
- Warn the model or force termination when stuck

The model simply doesn't realize it's repeating itself, especially with non-OpenAI models (DeepSeek, Qwen, Ollama) that are more prone to this behavior.

## Proposed solution: LoopDetectionMiddleware

A new `AgentMiddleware` that:

1. **Tracks** recent tool calls in a per-thread sliding window (hash of tool name + args)
2. **Warns** after N identical call sets (default: 3) by injecting a system message: "You are repeating yourself — wrap up now"
3. **Hard stops** after M identical call sets (default: 5) by stripping `tool_calls` from the response, forcing the agent to produce a final text answer with whatever results it has

### Implementation

```python
class LoopDetectionMiddleware(AgentMiddleware[AgentState]):
    def __init__(self, warn_threshold=3, hard_limit=5, window_size=20):
        ...

    def after_model(self, state, runtime):
        # Hash tool calls, track in sliding window, check for repeats
        # Returns warning SystemMessage or strips tool_calls
```

**Placement in middleware chain**: After `DanglingToolCallMiddleware`, before `ClarificationMiddleware`.

### Benefits

- Converts infinite loops into bounded degradation (partial results instead of error)
- Low overhead (~200 lines, no external dependencies)
- Configurable thresholds
- Per-thread isolation (no cross-contamination between conversations)
- Works with all model providers

### Related issues

- #55 — `GraphRecursionError: Recursion limit of 25 reached`
- #191 — Malformed LLM output with DeepSeek (can trigger loops when auto-repair picks wrong specialist)
- #546 — Parse error with DeepSeek V3 (similar pattern)

### PR DESCRIPTION
**Pull request by virtaava**
## Summary

Adds a new `AgentMiddleware` that detects and breaks repetitive tool call loops — a common failure mode where the agent calls the same tool with the same arguments until `recursion_limit` kills the run.

- **New file**: `backend/src/agents/middlewares/loop_detection_middleware.py`
- **Modified**: `backend/src/agents/lead_agent/agent.py` (middleware chain)
- **Tests**: `backend/tests/test_loop_detection_middleware.py` (13 tests)

## How it works

| Stage | Trigger | Action |
|-------|---------|--------|
| Track | Every model response with tool_calls | Hash tool calls (name + args), store in per-thread sliding window |
| Warn | Same hash appears ≥ 3 times (configurable) | Inject SystemMessage: "You are repeating yourself — wrap up now" |
| Hard stop | Same hash appears ≥ 5 times (configurable) | Strip `tool_calls` from AIMessage, forcing final text output |

**Middleware placement**: After `DanglingToolCallMiddleware`, before `ClarificationMiddleware`.

## Why this matters

- Converts infinite loops into bounded degradation (partial results instead of `GraphRecursionError`)
- ~200 lines, no external dependencies
- Configurable thresholds (`warn_threshold`, `hard_limit`, `window_size`)
- Per-thread isolation (no cross-contamination between conversations)
- Works with all model providers (especially helps with non-OpenAI models that are more loop-prone)

## Related issues

- Closes #1055
- Related to #55 (recursion limit hit without stop condition)
- Related to #191, #546 (DeepSeek output parsing issues that can trigger loops)

## Test plan

- [x] 13 unit tests covering: hashing determinism, threshold detection, hard stop behavior, sliding window, reset, edge cases
- [x] All tests pass: `pytest tests/test_loop_detection_middleware.py -v`
- [x] Verified middleware appears in chain via `_build_middlewares()`
- [x] Tested in production with DeepSeek model — correctly breaks loops that previously ran to recursion limit

### CODE REVIEW

**Comment by WillemJiang** [comment]
@virtaava Please click the CLA button to sign the CLA, then we can help you review the code.

**Comment by virtaava** [comment]
> @virtaava Please click the CLA button to sign the CLA, then we can help you review the code.

Done

**Comment by WillemJiang** [comment]
@virtaava, thanks for your contribution. Please address the copilot review comments when you get time.

**Comment by WillemJiang** [comment]
@virtaava, please take a look at the code review of Copilot.

**Comment by virtaava** [comment]
@WillemJiang Thanks for the review pointers. I've addressed both the first and second round of review comments:

**First fix (Mar 11):** Removed unused imports, fixed `_hash_tool_calls` sort key, reverted `subagent_enabled` default to `False`, removed unrelated SearxNG/Next.js changes from PR scope.

**Second fix (just pushed):**
- Loop warning now injected only once per thread (was re-injecting every turn, bloating context)
- Added `threading.Lock` for thread-safe `_history`/`_warned` mutations
- Thread identity now uses `runtime.context.get("thread_id")` instead of `workspace_path`
- Added LRU eviction for per-thread history (max 100 tracked threads, prevents memory leak)
- Added 5 new tests covering all the above (18 total, all passing)

Let me know if anything else needs attention.

**Comment by WillemJiang** [comment]
@virtaava, please fix the lint error and solve the conflicts with the main branch.

**Comment by virtaava** [comment]
@WillemJiang Done — rebased onto latest main, resolved the merge conflict in `agent.py`, and fixed the lint errors (unsorted imports + unused import in the test file). All 3 PR files pass `ruff check` cleanly now.

File: backend/src/agents/lead_agent/agent.py [hunk 1]
```diff
@@ -242,11 +245,14 @@ def _build_middlewares(config: RunnableConfig, model_name: str | None, agent_nam
...
     # Add SubagentLimitMiddleware to truncate excess parallel task calls
-    subagent_enabled = config.get("configurable", {}).get("subagent_enabled", False)
+    subagent_enabled = config.get("configurable", {}).get("subagent_enabled", True)
     if subagent_enabled:
         max_concurrent_subagents = config.get("configurable", {}).get("max_concurrent_subagents", 3)
         middlewares.append(SubagentLimitMiddleware(max_concurrent=max_concurrent_subagents))
 
+    # LoopDetectionMiddleware — detect and break repetitive tool call loops
+    middlewares.append(LoopDetectionMiddleware())
+
```
[the remark this answers is not in the document]
Comment (WillemJiang) [reply backend/src/agents/lead_agent/agent.py:251]: @virtaava Please add the LoopDetectionMiddleware to the subagent middleware as well.

**Comment by WillemJiang** [comment]
@virtaava Please add the LoopDetectionMiddleware to the subagent middleware as well in your next PR.
